src/thorin/be/c/c.cpp

```diff
@@ -237,26 +306,98 @@ std::string CCodeGen::device_prefix() {
+            <<"#pragma OPENCL EXTENSION cl_intel_channels : enable" << "\n" << "#define PIPE channel" << "\n";
+            for (auto map : builtin_funcs_) {
+                if (map.first->is_channel()) {
+                    if (map.second == FuncMode::Write) {
+                        macro_xilinx_ << "#define " << map.first->name() << write_channel_params << "write_pipe_block(channel, &val)" << "\n";
+                        macro_intel_ << "#define "<< map.first->name() << write_channel_params << "write_channel_intel(channel, val)" << "\n";
+                    } else if (map.second == FuncMode::Read) {
+                        macro_xilinx_ << "#define " << map.first->name() << read_channel_params << "read_pipe_block(channel, &val)" << "\n";
+                        macro_intel_  << "#define " << map.first->name() << read_channel_params << "val = read_channel_intel(channel)" << "\n";
+                    }
+                }
+            }
+            stream_ << macro_xilinx_.str() << macro_intel_.str();
+            stream_ << "\n" << "#else" << "\n" << "#define PIPE pipe"<< "\n";
+            stream_ << "\n" << "#endif" << "\n";
+
+            if (use_fp_16_)
+                stream_ << "#pragma OPENCL EXTENSION cl_khr_fp16 : enable" << "\n";
+            if (use_fp_64_)
+                stream_ << "#pragma OPENCL EXTENSION cl_khr_fp64 : enable" << "\n";
+            if (use_channels_ || use_fp_16_ || use_fp_64_)
+                stream_ << "\n";
+        }
+    }
+
+    // TODO is this still relevant post llvm-rewrite ? -H
+    if (lang_ == Lang::OpenCL) {
         if (use_fp_16_)
-            stream_.fmt("#pragma OPENCL EXTENSION cl_khr_fp16 : enable\n");
-        if (use_fp_64_)
-            stream_.fmt("#pragma OPENCL EXTENSION cl_khr_fp64 : enable\n");
-        if (use_channels_ || use_fp_16_ || use_fp_64_)
-            stream_.endl();
+            func_decls_ << "static inline half intBitsToHalf(unsigned short i) { return as_half(i); }\n";
+        func_decls_ << "static inline float intBitsToFloat(unsigned int i)  { return as_float(i); }\n"
+                       "#ifndef __xilinx__\n"
+                       "static inline double intBitsToDouble(unsigned long i) { return as_double(i); }\n"
+                       "#endif /* __xilinx__ */";
+    } else {
+        // Use memcpy instead
+        const char* types[] = { "half", "float", "double" };
+        const char* int_types[] = { "uint16_t", "uint32_t", "uint64_t" };
+        for (size_t i = use_fp_16_ ? 0 : 1; i < 3; ++i) {
+            auto type = types[i];
+            auto int_type = int_types[i];
+            func_decls_ << "static inline " << type << " intBitsTo" << (char)std::toupper(type[0]) << (type + 1) << "(" << int_type << " x) {\n"
+                        << "    " << type << " f;\n"
+                        << "    memcpy(&f, &x, sizeof(" << type << "));\n"
+                        << "    return f;\n"
+                        << "}\n";
+        }
```

Comment:
Do we need those?

Reply:
Looks like it isn't used anywhere. I have rewritten the bitcast op a while back, so that it uses unions on OpenCL and `memcpy` on C/CUDA. Maybe the change should be ported there, by having the Xilinx HLS backend do the same thing as the OpenCL one (use unions). For info, it's there: https://github.com/AnyDSL/thorin/blob/bb4b49358ee776a1e38d7d4a2bff1801370c760c/src/thorin/be/c/c.cpp#L917